### DISCUSSION TITLE
TypeInitializerException

### DISCUSSION THREAD

**Discussion by TranstecJasonD** [NetTopologySuite/NetTopologySuite#535] [Q&A] [answered]
Hoping someone has seen this before. I was seeing this in my real app so I created a sample app and it has the same issue. No matter what I do, anything involving a factory is giving me:
"The type initializer for 'NetTopologySuite.Geometries.GeometryFactory' threw an exception."

Everything seems to work, but objects returned by a geometry that require calling the factory all give me this message. I've tried 2.2 and 2.3, VB and C#, setting the factory instance, not setting it, different precision models. Is this just a debugger issue? The Visual Studio Locals window is full of exceptions but if I just do something like the following, it shows up fine in the Output window:
System.Diagnostics.Debug.WriteLine(loRing.Centroid.ToString());

This is all I'm doing:
```csharp

private void CreateInt32() {
         Coordinate[] loCoordinates;
         GeometryFactory loFactory;

         loFactory = NtsGeometryServices.Instance.CreateGeometryFactory();

         loCoordinates = new Coordinate[5];
         loCoordinates[0] = new Coordinate(1, 1);
         loCoordinates[1] = new Coordinate(1, 3);
         loCoordinates[2] = new Coordinate(3, 3);
         loCoordinates[3] = new Coordinate(3, 1);
         loCoordinates[4] = new Coordinate(1, 1);
         var loRing = loFactory.CreateLinearRing(loCoordinates);
         System.Diagnostics.Debug.WriteLine(loRing.Centroid.ToString());
      }
```

**Comment by airbreather** [comment]
Can you please paste the `.csproj` file from your sample app?

**Comment by TranstecJasonD** [comment]
[WindowsFormsApp1.txt](https://github.com/NetTopologySuite/NetTopologySuite/files/6823649/WindowsFormsApp1.txt)

My real app is WinForms, so that's what my sample is.

**Comment by FObermaier** [comment]
It seems that  when using our netStandard2.0 NetTopologySuite package with a NetFramework project, NetTopologySuite gets wrapped using some sort of COM(+) stuff. 
* Can you change your project to target NETCORAPP3.1 or NET50? 
* Should we provide a NET48 library in our nuget package?

**Reply by DGuidi** [reply]
Unfortunately I have no time to do a test right now, but I used often NTS in netstandard format with net framework projects in the past months (even 4.6.1). Maybe I used older version (2.1.0 probably) because I don't remember that I used `NtsGeometryServices.Instance.CreateGeometryFactory()` to create a factory, but a simple `GeometryFactory.Default`...

**Reply by airbreather** [reply]
> Should we provide a NET48 library in our nuget package?

I maintain a .NET Framework 4.8 WPF application at work that is able to use NetTopologySuite 2.3.0 via `PackageReference` perfectly fine without any apparent weird COM stuff.

> I don't remember that I used `NtsGeometryServices.Instance.CreateGeometryFactory()` to create a factory, but a simple `GeometryFactory.Default`...

`GeometryFactory.Default` will work, but:

- `SRID` will always be 0
- `PrecisionModel` will always be floating double-precision
- `CoordinateSequenceFactory`, `GeometryOverlay`, and `CoordinateEqualityComparer` will be whatever was used for `NtsGeometryServices.Instance` at the time that `GeometryFactory`'s type initializer was run, which may or may not have happened before the user had the chance to override `NtsGeometryServices.Instance` as appropriate for their application

With `someNtsGeometryServicesObject.CreateGeometryFactory()`, users have more control and certainty over what values are used for the `GeometryFactory`.

**Reply by FObermaier** [reply, marked as answer]
To be clear, the sample WinForms application that I built according to the provided sample, works fine, too, unless you start to set breakpoints and investigate `Geometry` objects in the debugger's watch window. And this only happens if it targets some full .Net Framework.

**Reply by TranstecJasonD** [reply]
I can't change my app to Core or Net5, as much as I would like to. As mentioned, it does seem to actually work unless breakpoints are used. Which makes debugging kind of frustrating, but I can live with it. The only reason I went down this path is because I was experimenting with PrecisionModel and I didn't realize when I first started using this library that I should be using Factory stuff.
